### ISSUE TITLE
Entk on Polaris has issue that tasks running in parallel uses overlapping CPU resources

### ISSUE THREAD

**Issue by GKNB** [radical-cybertools/radical.entk#648]
The issue we see is that, when multiple tasks are submitted to the same stage on Polaris, those tasks are actually scheduled to use overlapping CPU cores instead of non-overlapping CPU cores. Below is an example:

```
#!/usr/bin/env python3

import radical.pilot as rp
import radical.entk  as re

import radical.entk.exceptions as ree


if __name__ == '__main__':
    task1 = re.Task({
            'executable': '/home/twang3/misc/example/GettingStarted/Examples/Theta/affinity/hello_affinity',
            'pre_exec'  : ['module load conda'],
            'cpu_reqs'  : {'cpu_processes'  : 2,
                           'cpu_threads'    : 4,
                           'cpu_thread_type': rp.OpenMP}
        })
    task2 = re.Task({
            'executable': '/home/twang3/misc/example/GettingStarted/Examples/Theta/affinity/hello_affinity',
            'pre_exec'  : ['module load conda'],
            'cpu_reqs'  : {'cpu_processes'  : 3,
                           'cpu_threads'    : 2,
                           'cpu_thread_type': rp.OpenMP}
        })
    task3 = re.Task({
            'executable': '/home/twang3/misc/example/GettingStarted/Examples/Theta/affinity/hello_affinity',
            'pre_exec'  : ['module load conda'],
            'cpu_reqs'  : {'cpu_processes'  : 4,
                           'cpu_threads'    : 1,
                           'cpu_thread_type': rp.OpenMP}
        })

    stage = re.Stage()
    stage.add_tasks([task1, task2, task3])

    pipeline = re.Pipeline()
    pipeline.add_stages([stage])
    try:
        amgr = re.AppManager()
        amgr.resource_desc = {
            'resource': 'anl.polaris',
            'project' : 'RECUP',
            'queue'   : 'debug',
            'cpus'    : 32,
            'gpus'    : 0,
            'walltime': 5
        }
        amgr.workflow = [pipeline]
        amgr.run()
    except (KeyboardInterrupt, ree.EnTKError):
        pass
```

Here, the executable we used are from 

https://github.com/argonne-lcf/GettingStarted/blob/master/Examples/Theta/affinity/main.cpp

In addition to that, at line 44 I also added `sleep(30);` in order to make sure that those three tasks are running in parallel. 

That output shows the following:

twang3@polaris-login-01:~> cat ~/radical.pilot.sandbox/re.session.polaris-login-01.twang3.019544.0001/pilot.0000/task.000000/task.000000.out 
To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 1  tid= 0: list_cores= (4-7)
To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 1  tid= 1: list_cores= (4-7)
To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 1  tid= 2: list_cores= (4-7)
To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 1  tid= 3: list_cores= (4-7)

To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 0  tid= 0: list_cores= (0-3)
To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 0  tid= 1: list_cores= (0-3)
To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 0  tid= 2: list_cores= (0-3)
To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 0  tid= 3: list_cores= (0-3)

twang3@polaris-login-01:~> cat ~/radical.pilot.sandbox/re.session.polaris-login-01.twang3.019544.0001/pilot.0000/task.000001/task.000001.out 
To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 2  tid= 0: list_cores= (4,5)
To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 2  tid= 1: list_cores= (4,5)

To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 0  tid= 0: list_cores= (0,1)
To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 0  tid= 1: list_cores= (0,1)

To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 1  tid= 0: list_cores= (2,3)
To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 1  tid= 1: list_cores= (2,3)

twang3@polaris-login-01:~> cat ~/radical.pilot.sandbox/re.session.polaris-login-01.twang3.019544.0001/pilot.0000/task.000002/task.000002.out 
To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 1  tid= 0: list_cores= (1)

To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 3  tid= 0: list_cores= (3)

To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 2  tid= 0: list_cores= (2)

To affinity and beyond!! nname= x3004c0s1b1n0  rnk= 0  tid= 0: list_cores= (0)

Suggesting that all of those three tasks are actually using overlapping CPU resources. 

**Updates:**

Mikhail suggests that on Polaris, instead of using the mpiexec provided by the default module, we can switch to a different mpirun which is provided by mvapich2. With his suggestions, I added the following to the ~/.radical/pilot/configs/resource_anl.json file for Polaris:

```
        "launch_methods"              : {
                                         "order" : ["MPIRUN"],
                                         "MPIRUN": {}
                                        },
        "pre_bootstrap_0"             : [
                                         "module load conda",
                                         "module load e4s/22.08/mvapich2"
                                        ]
```

and run the same script again. The results does not seem reasonable, which could be due to the fact that the executable is specific to cray. However, the launch.sh file also has some problem. For example, I find the following schedule command in task.000000/task.000000.launch.sh:

```
# ------------------------------------------------------------------------------
# launch commands
$RP_PROF launch_submit
( \
  /soft/ecp/ParaTools/opt/mvapich2/gdr/2.3.7/no-mpittool/no-openacc/cuda11.5/crayofed5.1/mpirun/gnu11.2.0/bin/mpirun  -np 2   -host x3008c0s7b1n0,x3008c0s7b1n0 $RP_TASK_SANDBOX/task.000000.exec.sh \
) 1> $RP_TASK_SANDBOX/task.000000.out \
  2> $RP_TASK_SANDBOX/task.000000.err
RP_RET=$?
$RP_PROF launch_collect
```

As we can see, this command only specify that we need to launch two processes, but it does not specify how many CPU cores should be pinned to each process (which should be four in this case), so I think there might also be something wrong with mvapich2.

Thanks!

**Comment by mtitov** [comment]
@GKNB @okilic1 please try the following branch with fixed MPIEXEC LM (also, this user-defined config file `~/.radical/pilot/configs/resource_anl.json` can be removed, since we'll use `mpiexec`)

```
pip uninstall radical.pilot -y
pip install git+https://github.com/radical-cybertools/radical.pilot.git@fix/lm_mpiexec
```

p.s. related PR https://github.com/radical-cybertools/radical.pilot/pull/2979

**Comment by mtitov** [comment]
@GKNB @okilic1 BTW, had the following observation - mo matter is there oversubscription or not, for all tasks execution time stays `30s` (as what we set explicitly with time sleep)
```
cat ~/radical.pilot.sandbox/<sid>/pilot.0000/task.00000*/task.00000*.prof | grep rank_start
cat ~/radical.pilot.sandbox/<sid>/pilot.0000/task.00000*/task.00000*.prof | grep rank_stop
```

**Comment by mtitov** [comment]
Resolved with PR mentioned above (merged into RP devel branch)
@GKNB Tianle, can you please add description of the issue you had (regarding `ppn` option) into the following ticket https://github.com/radical-cybertools/radical.pilot/issues/3014

Current ticket is closed
